src/machine.rs

```diff
@@ -1396,14 +1396,22 @@ impl<'tcx> Machine<'tcx> for MiriMachine<'tcx> {
                 .emit_diagnostic(NonHaltingDiagnostic::AccessedAlloc(alloc_id, AccessKind::Read));
         }
         // The order of checks is deliberate, to prefer reporting a data race over a borrow tracker error.
-        if let Some(genmc_ctx) = machine.data_race.as_genmc_ref() {
-            // We throw away the value we get from GenMC for non-atomic accesses, these are still handled by Miri
-            let _value = genmc_ctx.memory_load(machine, ptr.addr(), range.size)?;
-        } else if let AllocDataRaceHandler::Vclocks(data_race, weak_memory) = &alloc_extra.data_race
-        {
-            data_race.read(alloc_id, range, NaReadType::Read, None, machine)?;
-            if let Some(weak_memory) = weak_memory {
-                weak_memory.memory_accessed(range, machine.data_race.as_data_race_ref().unwrap());
+        match &machine.data_race {
+            GlobalDataRaceHandler::None => {}
+            GlobalDataRaceHandler::Genmc(genmc_ctx) => {
+                // We throw away the value we get from GenMC for non-atomic accesses, these are still handled by Miri
+                let _value = genmc_ctx.memory_load(machine, ptr.addr(), range.size)?;
```

Comment:
There's no reason the genmc wrapper should even return the value -- please discard it inside the wrapper.